src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -729,4 +729,11 @@ object messages {
            | - auxiliary constructors specify the implicit value
            |"""
   }
+
+  case class NoClassOrObjectAfterCase()(implicit ctx: Context)
+  extends Message(27) {
+    val kind = "Syntax"
+    val msg = hl"Only ${"class"} or ${"object"} are allowed after ${"case"}."
+    val explanation = ""
```

Comment:
Here are a few possible scenarios, starting with the one you mentioned:

1. next token is `TRAIT`, give explanation on case traits not being a thing
2. next token is `t' such that `t >= CHARLIT  && t =< IDENTIFIER` you can't have `case` followed by these outside a match expression
3. same for `TRUE` and `FALSE`

Have a look at [Tokens.scala](https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/parsing/Tokens.scala) for more tokens that could follow. This is somewhat open ended, so feel free to limit yourself to a few scenarios and  we'll detail this in #1589 :)